### PR TITLE
[DO NOT MERGE] Publishing API v2 test helper bugfixes and improvements

### PR DESCRIPTION
**Pull request by benilovj**
During work by the Core Team on the Publishing API migration, we've found some issues and areas of improvement in the Publishing API v2 test helpers. Since this PR includes some renames, it'll probably be a breaking version change.

This PR is best reviewed commit-by-commit.

To do:

- [ ] get broad consensus about the changes here
- [x] make sure that the asserts work against Whitehall before merging

/cc @gpeng @alphagov/team-gov-uk-core-formats-and-publishing

### CODE REVIEW

**Comment by gpeng** [comment]
This did break some tests in Whitehall but I've fixed them in [this PR](https://github.com/alphagov/whitehall/pull/2429).

File: lib/gds_api/test_helpers/publishing_api_v2.rb [hunk 1]
```diff
@@ -108,17 +119,18 @@ def assert_publishing_api(verb, url, attributes_or_matcher = {}, times = 1)
         end
       end
 
-      def request_json_matching(required_attributes)
+      def request_json_includes(required_attributes)
         ->(request) do
           data = JSON.parse(request.body)
-          required_attributes.to_a.all? { |key, value| data[key.to_s] == value }
+          deep_stringify_keys(required_attributes).
+            to_a.all? { |key, value| data[key.to_s] == value }
```
Comment (boffbowsh) [inline lib/gds_api/test_helpers/publishing_api_v2.rb]: redundant `.to_s`

File: lib/gds_api/test_helpers/publishing_api_v2.rb [hunk 2]
```diff
@@ -150,6 +162,23 @@ def stub_publishing_api_put(content_id, body, resource_path, override_response_h
...
+      end
+
+      def deep_transform_keys(object, &block)
+        case object
+        when Hash
+          object.each_with_object({}) do |(key, value), result|
+            result[yield(key)] = deep_transform_keys(value, &block)
+          end
+        when Array
+          object.map{|item| deep_transform_keys(item, &block) }
```
Comment (benlovell) [inline lib/gds_api/test_helpers/publishing_api_v2.rb]: Missing a space here.

File: lib/gds_api/test_helpers/publishing_api_v2.rb (hunk 1 above)
Comment (gpeng) [inline lib/gds_api/test_helpers/publishing_api_v2.rb]: Fixed in 3d3b847

File: lib/gds_api/test_helpers/publishing_api_v2.rb (hunk 2 above)
Comment (gpeng) [inline lib/gds_api/test_helpers/publishing_api_v2.rb]: Fixed in 3d3b847

**Pull request merged**